fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java

```diff
@@ -0,0 +1,411 @@
+        for (Pair<Integer, Integer> slotPair : newSlotsArray) {
+           Pair<Expr, Expr> pair = new Pair<>(integerExprMap.get(slotPair.first), integerExprMap.get(slotPair.second));
+           Pair<Expr, Expr> eqPair = new Pair<>(integerExprMap.get(slotPair.second), integerExprMap.get(slotPair.first));
+           if (!slotEqSlotPredicatesSet.contains(pair)
+                && !slotEqSlotPredicatesSet.contains(eqPair)) {
+               slotEqSlotPredicatesSet.add(pair);
+               slotEqSlotPredicatesList.add(
+                       new BinaryPredicate(BinaryPredicate.Operator.EQ, integerExprMap.get(slotPair.first), integerExprMap.get(slotPair.second)));
+           }
+        }
+    }
+
+    private void inferSlotPredicates(ArrayList<Expr> slotEqSlotPredicatesList,
+                                     Set<Pair<Expr, Expr>> slotToLiteralPredicatesSet,
+                                     ArrayList<Expr> slotToLiteralPredicatesList,
+                                     ArrayList<Pair<Expr, Boolean>> newPredicatesList) {
+        for (Expr slotToLiteral : slotToLiteralPredicatesList) {
+            buildNewBinaryPredicate(slotToLiteral, slotEqSlotPredicatesList,
+                    slotToLiteralPredicatesSet, newPredicatesList);
+        }
+    }
+
+    /**
+     * Traverse slotEqSlot to extract slots with equal expressions and construct a new slotToLiteral.
+     */
+    private void buildNewBinaryPredicate(Expr slotToLiteral,
+                                         ArrayList<Expr> slotEqSlotPredicatesList,
+                                         Set<Pair<Expr, Expr>> slotToLiteralPredicatesSet,
+                                         ArrayList<Pair<Expr, Boolean>> newPredicatesList) {
+        SlotRef checkSlot = slotToLiteral.getChild(0).unwrapSlotRef();
+        for (Expr conjunct : slotEqSlotPredicatesList) {
+            SlotRef leftSlot = conjunct.getChild(0).unwrapSlotRef();
+            SlotRef rightSlot = conjunct.getChild(1).unwrapSlotRef();
+
+            if (checkSlot.simple_equals(leftSlot)) {
+                addNewBinaryPredicate(genNewBinaryPredicate(slotToLiteral, rightSlot),
+                        slotToLiteralPredicatesSet, newPredicatesList,
+                        isNeedInfer(rightSlot, leftSlot));
+            } else if (checkSlot.simple_equals(rightSlot)) {
+                addNewBinaryPredicate(genNewBinaryPredicate(slotToLiteral, leftSlot),
+                        slotToLiteralPredicatesSet, newPredicatesList,
+                        isNeedInfer(leftSlot, rightSlot));
+            }
+        }
+    }
+
+    /**
+     * To determine whether it needs to be extended.
+     * eg:t1.id = t2.id and t2.id = 1;
+     * @param newSlot: t1.id
+     * @param checkSlot: t2.id
+     * @return needInfer.    True: needInfer. False: not needInfer
+     */
+    private boolean isNeedInfer(SlotRef newSlot, SlotRef checkSlot) {
```

Comment:
FROM t0 left outer join t1 inner join t2 where t1.id=t2.id and t2.id=1
jop = left outer join
checkjop = inner join
But it still can infer , right?

Reply:
done